src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/CSharp/RemappingCSharpCodeActionResolver.cs

```diff
@@ -0,0 +1,142 @@
+
+        public async override Task<CodeAction> ResolveAsync(
+            CSharpCodeActionParams csharpParams,
+            CodeAction codeAction,
+            CancellationToken cancellationToken)
+        {
+            if (csharpParams is null)
+            {
+                throw new ArgumentNullException(nameof(csharpParams));
+            }
+
+            if (codeAction is null)
+            {
+                throw new ArgumentNullException(nameof(codeAction));
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var resolvedCodeAction = await ResolveCodeActionWithServerAsync(codeAction, cancellationToken).ConfigureAwait(false);
+            if (resolvedCodeAction.Edit?.DocumentChanges is null)
+            {
+                // Unable to resolve code action with server, return original code action
+                return codeAction;
+            }
+
+            if (resolvedCodeAction.Edit.DocumentChanges.Count() != 1)
+            {
+                // We don't yet support multi-document code actions, return original code action
+                return codeAction;
+            }
+
+            var documentChanged = resolvedCodeAction.Edit.DocumentChanges.First();
+            if (!documentChanged.IsTextDocumentEdit)
+            {
+                // Only Text Document Edit changes are supported currently, return original code action
+                return codeAction;
+            }
+
+            var textEdit = documentChanged.TextDocumentEdit.Edits.FirstOrDefault();
+            if (textEdit is null)
+            {
+                // No text edit available
+                return codeAction;
+            }
+
+            var documentSnapshot = await Task.Factory.StartNew(() =>
+            {
+                _documentResolver.TryResolveDocument(csharpParams.RazorFileUri.ToUri().GetAbsoluteOrUNCPath(), out var documentSnapshot);
+                return documentSnapshot;
+            }, cancellationToken, TaskCreationOptions.None, _foregroundDispatcher.ForegroundScheduler).ConfigureAwait(false);
+
+            if (documentSnapshot is null)
+            {
+                return codeAction;
+            }
+
+            var codeDocument = await documentSnapshot.GetGeneratedOutputAsync().ConfigureAwait(false);
+            if (codeDocument.IsUnsupported())
+            {
+                return codeAction;
+            }
+
+            if (!_documentMappingService.TryMapFromProjectedDocumentRange(
+                    codeDocument,
+                    textEdit.Range,
+                    out var originalRange))
+            {
+                // Text edit failed to map
+                return codeAction;
+            }
+
+            textEdit.Range = originalRange;
+
+            var documentVersion = await Task.Factory.StartNew(() =>
+            {
+                _documentVersionCache.TryGetDocumentVersion(documentSnapshot, out var version);
```

Comment:
I'd just do this in the original foreground dispatcher jump. It isn't super cheap to be jumping to & from the foreground thread so it'd be better to just do it once if we can.